File: algorithms/04_java/sorting/QuickSort.java
public class QuickSort {
    public static void sort(int[] a, int lo, int hi){
        if(lo<hi){int p=partition(a,lo,hi); sort(a,lo,p-1); sort(a,p+1,hi);}
    }
    static int partition(int[] a,int lo,int hi){
        int pivot=a[hi],i=lo-1;
        for(int j=lo;j<hi;j++) if(a[j]<=pivot){i++;int t=a[i];a[i]=a[j];a[j]=t;}
        int t=a[i+1];a[i+1]=a[hi];a[hi]=t; return i+1;
    }
}

File: algorithms/21_scala/sorting/QuickSort.scala
object QuickSort {
  def sort[T](arr: List[T])(implicit ord: Ordering[T]): List[T] = arr match {
    case Nil => Nil
    case pivot :: tail =>
      val (less, greater) = tail.partition(ord.lt(_, pivot))
      sort(less) ::: pivot :: sort(greater)
  }

  def main(args: Array[String]): Unit = {
    val data = List(33, 7, 91, 12, 5, 5, 78, 2, 44, 19)
    val sorted = sort(data)
    assert(sorted == data.sorted, "not sorted")
    println(s"[Scala QuickSort] Functional quicksort verified: $sorted")
  }
}
